src/ucp/wireup/select.c

```diff
@@ -995,21 +1093,27 @@ static ucs_status_t ucp_wireup_add_bw_lanes(ucp_ep_h ep, unsigned address_count,
     ucp_md_map_t md_map;
     int is_proxy;
 
-    num_lanes          = 0;
-    md_map             = bw_info->md_map;
-    local_dev_bitmap   = bw_info->local_dev_bitmap;
-    remote_dev_bitmap  = bw_info->remote_dev_bitmap;
+    num_lanes         = 0;
+    md_map            = bw_info->md_map;
+    local_dev_bitmap  = bw_info->local_dev_bitmap;
+    remote_dev_bitmap = bw_info->remote_dev_bitmap;
 
     /* lookup for requested number of lanes or limit of MD map
      * (we have to limit MD's number to avoid malloc in
      * memory registration) */
     while ((num_lanes < bw_info->max_lanes) &&
            (ucs_popcount(md_map) < UCP_MAX_OP_MDS)) {
-        status = ucp_wireup_select_transport(ep, address_list, address_count,
-                                             &bw_info->criteria, tl_bitmap, -1,
-                                             local_dev_bitmap, remote_dev_bitmap,
-                                             0, &select_info);
-        if (status != UCS_OK) {
+        status = ucp_wireup_select_lane(ep, address_count, address_list,
+                                        &bw_info->criteria, tl_bitmap, -1,
+                                        local_dev_bitmap, remote_dev_bitmap,
+                                        0, &select_info);
+        /* Break if: */
+        /* - selection wasn't OK */
+        if ((status != UCS_OK) ||
+            /* - the selected transport isn't scalable enough, but
+             *   the selected AM transport provides better scalability */
+            !ucp_wireup_check_scalability(context, select_info.max_num_eps,
```

Comment:
no need to compare with AM lane, just do not add lane if it is not scalable enough and am_emulation is allowed (otherwise we should not rely on scalability constraints )